### ISSUE TITLE
SsoAccessTokenProvider throws an ExpiredTokenException before the access token has expired

### ISSUE THREAD

**Issue by oscarostrand** [aws/aws-sdk-java-v2#3679]
### Describe the bug

When using SSO with the SDK the SsoAccessTokenProvider will throw an exception because of an expired token
15 min before the token actually expires. This seam to be intentional based on the code from the class:

SsoAccessTokenProvider.java

```Java
@SdkInternalApi
public final class SsoAccessTokenProvider implements SdkTokenProvider {

    //Removed rest of code for readability
    
    private SdkToken getTokenFromJson(String json) {
        JsonNode jsonNode = PARSER.parse(json);
        String expiration = jsonNode.field("expiresAt").map(JsonNode::text).orElse(null);

        Validate.notNull(expiration,
                         "The SSO session's expiration time could not be determined. Please refresh your SSO session.");

        if (tokenIsInvalid(expiration)) {
            throw ExpiredTokenException.builder().message("The SSO session associated with this profile has expired or is"
                                                          + " otherwise invalid. To refresh this SSO session run aws sso"
                                                          + " login with the corresponding profile.").build();
        }

        return SsoAccessToken.builder()
                             .accessToken(jsonNode.asObject().get("accessToken").text())
                             .expiresAt(Instant.parse(expiration)).build();

    }

    private boolean tokenIsInvalid(String expirationTime) {
        return Instant.now().isAfter(Instant.parse(expirationTime).minus(15, MINUTES));
    }

}
```

According to the error message, all you need to do is run the "aws sso login" command again to get a new session. 
However, the login command will not refresh the token in the SSO cache unless it has actually expired. This means
that for the last 15 min before the access token expires SSO does not work with the SDK.

I wrote a workaround by fetching the access token myself from the SSO cache file and using that
to get the credentials from the SSOClient. But it would be good if this could be handled by the SDK.

### Expected Behavior

The SsoAccessTokenProvider should not throw an ExpiredTokenException before the token has expired.  

### Current Behavior

The SsoAccessTokenProvider throws an ExpiredTokenException 15 min before the access token has expired. It also 
gives an error suggesting a solution that will not work until after the token has actually expired.

### Reproduction Steps

The bug is hard to reproduce naturally because it only occurs 15 min before the SSO access token expires. But it can be manually created by:

1. run the "aws sso login"  command
2. Change the "expiresAt"  time for the access token in the SSO cache file (~/.aws/sso/cache/xxxxx.json) to a value in the 15 min time span that the error occurs (for example current time + 10 min).
3. Use any AWS client in the SDK to make a request in any project configured to use SSO, for example:
```Java

S3Client.create().listBuckets();

```

### Possible Solution

The most obvious solution seems to be to change the "tokenIsInvalid" method in the SsoAccessTokenProvider class so that it no longer subtracts the 15 min from the expiration time. However, because the subtractions seem so intentional it's highly possible that there is something I am missing. 

Another solution would be to change the AWS CLI to refresh the access token when running the "aws sso login" command
even if it has not expired. Then the solution in the error message would at least work.

### AWS Java SDK version used

2.19.4

### JDK version used

 JDK 17.0.4

### Operating System and version

MacOs Monterey 12.6

**Comment by yasminetalby** [comment]
Hello @oscarostrand ,

Thank you very much for your submission. 
It seems that this behavior has also been reported by user of other SDK (specifically v3 javascript SDK).

I digged a little into their documentation and it seems that they follow a similar behavior where there is a 15 minutes time window in which the SDK will treat the SSO token as expired before the defined expiration date in token ([see](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/modules/_aws_sdk_credential_provider_sso.html)).

Per the CLI documentation, each time the login command is called, a new SSO access token will be retrieved ([see](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/sso/login.html)).

However, it seems that you are not able to get a new SSO access token if the previous one is not expired (not in the 15 minute expiration window but actually expired).

There has been an issue open in the `aws-cli` repository regarding a similar behavior ([see](https://github.com/aws/aws-cli/issues/7361)). 

I was wondering which version of the AWS CLI you are using and if the pattern described by the AWS-CLI submission issue matches what you observe?

Thank you very much for your time and collaboration!

Sincerely,

Yasmine.

**Comment by oscarostrand** [comment]
Hello @yasminetalby.

Thank you for helping out!

I have verified the behavior with version 2.9.17 of the AWS CLI. And you are right, on closer inspection, I see that the problem matches the pattern of the issue you linked. I do get a new access token, but the expiration time is not updated. This in turn causes the problem with SDK.

**Comment by yasminetalby** [comment]
Hello @oscarostrand ,

Thank you very much for your update. I am currently investigating this issue and will bring this up to tim-finnigan. 

Thank you very much for bringing this up to our attention.

Sincerely,

Yasmine

**Comment by joshtkehoe** [comment]
We are having the same problem, except we are using the JavaScript SDK. Our engineers login to AWS with SSO in the morning, and then 7 hours and 45 minutes later, they are unable to use the SDK or refresh the token for 15 minutes until the `expiresAt` time elapses. At that time, they are able to log in with sso again and get a new token. The inability to refresh the `expiresAt` time paired with the token invalidating 15 minutes before this time creates a dead zone where our engineers cannot use the SDK locally at all.

**Comment by oscarostrand** [comment]
@joshtkehoe we solved it by adding our own credential provider at the end of the provider chain that will simply get the token even if it has expired. Because you get a new valid token when you run the login command (even if the expiration time is never updated) this solves the issue without messing with the original behavior. I think this is a valid workaround until the issue can be fixed.

**Comment by debora-ito** [comment]
The fix was merged, and released in SDK version `2.20.100`.

Please try the fix and let us know if you still see the issue with the expiration time.
